org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/ProblemTypeAndMethodTest.java

```diff
@@ -8828,4 +8812,1430 @@ public void testBug576735() {
+	Util.delete(new File(OUTPUT_DIR, "p1" + File.separator + "A2.class"));
+	runner.shouldFlushOutputDirectory = false;
+
+	runner.customOptions.put(CompilerOptions.OPTION_ReportVarargsArgumentNeedCast, CompilerOptions.IGNORE);
+	runner.testFiles = new String[] {
+			"p2/C.java",
+			"""
+			package p2;
+			import p1.B;
+			public class C {
+				void test(B b) {
+					new B(this); // simply ambiguous as we don't even look at A1 or A2
+					new B(this, null);
+					new B(this, b);
+					new B(this, b, b);
+					new B(this, new B[0]);
+				}
+			}
+			"""
+		};
+	runner.expectedCompilerLog = """
+			----------
+			1. ERROR in p2\\C.java (at line 5)
+				new B(this); // simply ambiguous as we don't even look at A1 or A2
+				^^^^^^^^^^^
+			The constructor B(Object, B[]) is ambiguous
+			----------
+			2. ERROR in p2\\C.java (at line 6)
+				new B(this, null);
+				^^^^^^^^^^^^^^^^^
+			The constructor B(Object, A1...) refers to the missing type A1
+			----------
+			3. ERROR in p2\\C.java (at line 7)
+				new B(this, b);
+				^^^^^^^^^^^^^^
+			The constructor B(Object, A1...) refers to the missing type A1
+			----------
+			4. ERROR in p2\\C.java (at line 8)
+				new B(this, b, b);
+				^^^^^^^^^^^^^^^^^
+			The constructor B(Object, A1...) refers to the missing type A1
+			----------
+			5. ERROR in p2\\C.java (at line 9)
+				new B(this, new B[0]);
+				^^^^^^^^^^^^^^^^^^^^^
+			The constructor B(Object, A1...) refers to the missing type A1
+			----------
+			""";
+	runner.runNegativeTest();
+}
+public void testMissingClassNeededForOverloadResolution_varargs1b() {
+	// like testMissingClassNeededForOverloadResolution_varargs1a, but no preceding regular parameter
+	if (this.complianceLevel < ClassFileConstants.JDK1_8) return; // ignore different outcome below 1.8 since PR 2543
+	Runner runner = new Runner();
+	runner.customOptions = getCompilerOptions();
+	runner.testFiles = new String[] {
+			"p1/A1.java",
+			"""
+			package p1;
+			public class A1 {}
+			""",
+			"p1/A2.java",
+			"""
+			package p1;
+			public class A2 {}
+			""",
+			"p1/B.java",
+			"""
+			package p1;
+			public class B {
+				public void m(String... s) {}
+				public void m(A1... a) {}
+				public void m(A2... a) {}
+			}
+			"""
+		};
+	runner.runConformTest();
+
+	// delete binary files A1, A1 (i.e. simulate removing it from classpath for subsequent compile)
+	Util.delete(new File(OUTPUT_DIR, "p1" + File.separator + "A1.class"));
+	Util.delete(new File(OUTPUT_DIR, "p1" + File.separator + "A2.class"));
+	runner.shouldFlushOutputDirectory = false;
+
+	runner.customOptions.put(CompilerOptions.OPTION_ReportVarargsArgumentNeedCast, CompilerOptions.IGNORE);
+	runner.testFiles = new String[] {
+			"p2/C.java",
+			"""
+			package p2;
+			import p1.B;
+			public class C {
+				void test(B b) {
+					b.m(); // simply ambiguous as we don't even look at A1 or A2
+					b.m(null);
+					b.m(b);
+					b.m(b, b);
+					b.m(new B[0]);
```

Comment:
Can we have some calls here which target `m(String... s)` just for variety and to show behavior ?

Reply:
Well, that will leave us again with "compatibility checks would like to see the missing type", even though more elaborate checks could make a safe decision. 

Actually one kind of invocation will pass: `b.m(new String[0])`, i.e., with exact match we won't even bother about other overloads.

Anyway, I'll add tests showing this sub-optimal behavior.